### PR TITLE
Enable more MSVC warnings for SDL signoff

### LINKED ISSUE #40525: SDL signoff requirements - please enable additional C/C++ compiler warnings
> __Please prioritize this work.__ Depending on the size of the code base, there could be hundreds or thousands of warnings which require addressing.
> 
> __This work must be completed before RC1 snap.__

Recent SDL changes require that we enable additional compiler warnings across our C/C++ projects. The requirements are:

* Compile with warning level 3 or higher. These warnings do not have to be fixed, but they must at least be triaged.
* Enable the following additional warnings, which __must__ be fixed if they occur:
  * __C4018__ - 'expression' : signed/unsigned mismatch
  * __C4055__ - 'conversion' : from data pointer 'type1' to function pointer 'type2'
  * __C4146__ - unary minus operator applied to unsigned type, result still unsigned
  * __C4242__ - 'identifier' : conversion from 'type1' to 'type2', possible loss of data
  * __C4244__ - 'conversion' conversion from 'type1' to 'type2', possible loss of data
  * __C4267__ - 'var' : conversion from 'size_t' to 'type', possible loss of data
  * __C4302__ - 'conversion' : truncation from 'type 1' to 'type 2'
  * __C4308__ - negative integral constant converted to unsigned type
  * __C4509__ - nonstandard extension used: 'function' uses SEH and 'object' has destructor
  * __C4510__ - 'class' : default constructor could not be generated
  * __C4532__ - 'continue' : jump out of __finally/finally block has undefined behavior during termination handling
  * __C4533__ - initialization of 'variable' is skipped by 'instruction'
  * __C4610__ - object 'class' can never be instantiated - user-defined constructor required
  * __C4611__ - interaction between 'function' and C++ object destruction is non-portable
  * __C4700__ - uninitialized local variable 'name' used
  * __C4701__ - Potentially uninitialized local variable 'name' used
  * __C4703__ - Potentially uninitialized local pointer variable 'name' used
  * __C4789__ - destination of memory copy is too small
  * __C4995__ - 'function': name was marked as #pragma deprecated
  * __C4996__ - 'function': was declared deprecated also 'std::': Function call with parameters that are potentially unsafe - this call relies on the caller to check that the passed values are correct. To disable this warning, use -D_SCL_SECURE_NO_WARNINGS. See documentation on how to use Visual C++ 'Checked Iterators'

See [the SDL guidelines](https://liquid.microsoft.com/Web/Object/Read/ms.security/Requirements/Microsoft.Security.SystemsADM.10086) (MSFT internal only) for further information.

To make auditing code easier prior to signoff and release, **please include the string _Microsoft.Security.SystemsADM.10086_** somewhere in the vcxproj / cmake / whatever file which is responsible for C/C++ compilation where all of these flags are defined. That gives the code auditors a target string to search against and validate that the proper warnings are enabled. See below an example of how to do this in a cmake file.

```cmake
# [[! Microsoft.Security.SystemsADM.10086 !]] - SQL required warnings
add_compile_options($<$<COMPILE_LANGUAGE:C,CXX>:/W3>) # warning level 3
add_compile_options($<$<COMPILE_LANGUAGE:C,CXX>:/WX>) # treat warnings as errors
add_compile_options($<$<COMPILE_LANGUAGE:C,CXX>:/we4018>) # 'expression' : signed/unsigned mismatch
add_compile_options($<$<COMPILE_LANGUAGE:C,CXX>:/we4055>) # 'conversion' : from data pointer 'type1' to function pointer 'type2'
# ... and so on ...
```

Once the work is finished, or if there is no work to perform, please feel free to close this issue.

Thanks for your assistance!

## Quick FAQ

__What code is bound to this requirement?__

This affects only _production_ code. Production code is generally defined as code which ships as part of the product and which runs on customer machines or which manages infrastructure, such as our build labs. Unit and functional test projects are not considered production code.

__Does this need to be backported?__

No backporting plans at this time. If actual bugs are found during this process, individual product teams have discretion to selectively backport into the next downlevel servicing vehicle.

__What about forked third-party code?__

This requirement applies to all code that MSFT builds from source, regardless of its provenance. Ideally any changes that we make to local forked copies can be submitted upstream as a PR so that the wider ecosystem can enjoy their benefits.

If this is impractical, exceptions to this requirement can be sought on an as-needed basis. However, exceptions are: (a) not guaranteed to be granted; and (b) time-constrained. The exception process is not intended to provide a permanent deferral of this work. Please contact the _fxsecurity_ alias if an exception is needed.

__What about C# and other languages?__

This requirement only affects C/C++ code. Requirements for other languages will be filed as separate issues.

### PR DESCRIPTION
**Pull request by adityamandaleeka**
Fix https://github.com/dotnet/aspnetcore/issues/40525

### CODE REVIEW

File: src/Servers/IIS/AspNetCoreModuleV2/IISLib/IISLib.vcxproj [hunk 1]
```diff
@@ -78,11 +78,17 @@
     <Import Project="$(UserRootDir)\Microsoft.Cpp.$(Platform).user.props" Condition="exists('$(UserRootDir)\Microsoft.Cpp.$(Platform).user.props')" Label="LocalAppDataPlatform" />
   </ImportGroup>
   <PropertyGroup Label="UserMacros" />
+  <ItemDefinitionGroup>
+    <ClCompile>
+      <!-- Microsoft.Security.SystemsADM.10086 - SDL required warnings -->
+      <WarningLevel>Level4</WarningLevel>
```
Comment (HaoK) [inline src/Servers/IIS/AspNetCoreModuleV2/IISLib/IISLib.vcxproj:84]: Can you try having IISLib import Build.Lib.Settings now after this PR I think you fixed all the main things that was blocking this when I was cleaning things up

**Comment by HaoK** [comment]
Do we also need to do this for the installer CAs? i.e. the vcxproj under https://github.com/dotnet/aspnetcore/tree/main/src/Installers/Windows

**Comment by adityamandaleeka** [comment]
> Do we also need to do this for the installer CAs? i.e. the vcxproj for aspnetca/iisca and the random other ones under https://github.com/dotnet/aspnetcore/tree/main/src/Installers/Windows

Good call, yea, I think so.

File: src/Servers/IIS/AspNetCoreModuleV2/IISLib/IISLib.vcxproj (hunk 1 above)
Comment (adityamandaleeka) [reply src/Servers/IIS/AspNetCoreModuleV2/IISLib/IISLib.vcxproj]: OK, done in my latest push. The shared settings have conformance mode on so I decided to fix those warnings too. So many unnecessary goto's 😆 . I want to do another cleanup pass on this but didn't want to make this PR too complex.

**Comment by adityamandaleeka** [comment]
I think we should keep these commits separate FWIW so we can easily separate the SDL warning changes and the IISLib cleanup if needed in the future.

File: src/Servers/IIS/AspNetCoreModuleV2/IISLib/IISLib.vcxproj [hunk 2]
```diff
@@ -1,123 +1,36 @@
...
-  <PropertyGroup Label="UserMacros" />
   <ItemDefinitionGroup Condition="'$(Configuration)|$(Platform)'=='Debug|Win32' OR'$(Configuration)|$(Platform)'=='Debug|x64' OR '$(Configuration)|$(Platform)'=='Debug|ARM64' OR '$(Configuration)|$(Platform)'=='Debug|Any CPU'">
     <ClCompile>
       <PrecompiledHeader>
       </PrecompiledHeader>
-      <WarningLevel>Level3</WarningLevel>
-      <Optimization>Disabled</Optimization>
-      <PreprocessorDefinitions>WIN32;NDEBUG;_LIB;%(PreprocessorDefinitions)</PreprocessorDefinitions>
+      <PreprocessorDefinitions>WIN32;%(PreprocessorDefinitions)</PreprocessorDefinitions>
       <SDLCheck>true</SDLCheck>
```
Comment (adityamandaleeka) [inline src/Servers/IIS/AspNetCoreModuleV2/IISLib/IISLib.vcxproj:17]: I left in settings that differ from the common set (like this one).

File: src/Servers/IIS/AspNetCoreModuleV2/IISLib/IISLib.vcxproj [hunk 3]
```diff
@@ -1,123 +1,36 @@
...
-    <CharacterSet>Unicode</CharacterSet>
-  </PropertyGroup>
-  <Import Project="$(VCTargetsPath)\Microsoft.Cpp.props" />
-  <ImportGroup Label="ExtensionSettings">
-  </ImportGroup>
-  <ImportGroup Label="PropertySheets">
-    <Import Project="$(UserRootDir)\Microsoft.Cpp.$(Platform).user.props" Condition="exists('$(UserRootDir)\Microsoft.Cpp.$(Platform).user.props')" Label="LocalAppDataPlatform" />
-  </ImportGroup>
-  <PropertyGroup Label="UserMacros" />
   <ItemDefinitionGroup Condition="'$(Configuration)|$(Platform)'=='Debug|Win32' OR'$(Configuration)|$(Platform)'=='Debug|x64' OR '$(Configuration)|$(Platform)'=='Debug|ARM64' OR '$(Configuration)|$(Platform)'=='Debug|Any CPU'">
```
Comment (HaoK) [inline src/Servers/IIS/AspNetCoreModuleV2/IISLib/IISLib.vcxproj:12]: nit: can we shorten this to just $(Configuration) == Debug since its all 4 platforms?

File: src/Servers/IIS/AspNetCoreModuleV2/IISLib/IISLib.vcxproj [hunk 4]
```diff
@@ -1,123 +1,36 @@
...
-      <ShowIncludes>false</ShowIncludes>
-      <TreatWarningAsError>true</TreatWarningAsError>
-      <MultiProcessorCompilation>true</MultiProcessorCompilation>
     </ClCompile>
     <Link>
-      <SubSystem>Windows</SubSystem>
       <GenerateDebugInformation>true</GenerateDebugInformation>
     </Link>
   </ItemDefinitionGroup>
   <ItemDefinitionGroup Condition="'$(Configuration)|$(Platform)'=='Release|Win32' OR '$(Configuration)|$(Platform)'=='Release|x64' OR '$(Configuration)|$(Platform)'=='Release|ARM64' OR '$(Configuration)|$(Platform)'=='Release|Any CPU'">
```
Comment (HaoK) [inline src/Servers/IIS/AspNetCoreModuleV2/IISLib/IISLib.vcxproj:24]: nit: similarly is this just Config = release?

File: src/Servers/IIS/AspNetCoreModuleV2/IISLib/stringa.cpp [hunk 5]
```diff
@@ -1166,28 +1169,23 @@ STRA::AuxAppendWTruncate(
...
     {
-        hr = HRESULT_FROM_WIN32( ERROR_ARITHMETIC_OVERFLOW );
-        goto Finished;
+        return HRESULT_FROM_WIN32( ERROR_ARITHMETIC_OVERFLOW );
     }
 
     if( !m_Buff.Resize( static_cast<SIZE_T>(cbNeeded) ) )
     {
-        hr = E_OUTOFMEMORY;
-        goto Finished;
```
Comment (HaoK) [inline src/Servers/IIS/AspNetCoreModuleV2/IISLib/stringa.cpp:1190]: Wow gotos!

**Review by HaoK — APPROVED** [review]

File: src/Servers/IIS/AspNetCoreModuleV2/IISLib/stringa.cpp (hunk 5 above)
Comment (GrabYourPitchforks) [reply src/Servers/IIS/AspNetCoreModuleV2/IISLib/stringa.cpp:1190]: Sometimes it's C code spiffed up in a .cpp file extension. :)
